src/py/flwr/client/client_app.py

```diff
@@ -54,25 +56,134 @@ class ClientApp:
 
     def __init__(
         self,
-        client_fn: ClientFn,  # Only for backward compatibility
+        client_fn: Optional[ClientFn] = None,  # Only for backward compatibility
         mods: Optional[List[Mod]] = None,
     ) -> None:
+        self._call: Optional[ClientAppCallable] = None
+
         # Create wrapper function for `handle`
-        def ffn(
-            message: Message,
-            context: Context,
-        ) -> Message:  # pylint: disable=invalid-name
-            out_message = handle_legacy_message_from_msgtype(
-                client_fn=client_fn, message=message, context=context
-            )
-            return out_message
-
-        # Wrap mods around the wrapped handle function
-        self._call = make_ffn(ffn, mods if mods is not None else [])
+        if client_fn is not None:
+
+            def ffn(
+                message: Message,
+                context: Context,
+            ) -> Message:  # pylint: disable=invalid-name
+                out_message = handle_legacy_message_from_msgtype(
+                    client_fn=client_fn, message=message, context=context
+                )
+                return out_message
+
+            # Wrap mods around the wrapped handle function
+            self._call = make_ffn(ffn, mods if mods is not None else [])
+
+        # Step functions
+        self._train: Optional[ClientAppCallable] = None
+        self._evaluate: Optional[ClientAppCallable] = None
+        self._query: Optional[ClientAppCallable] = None
 
     def __call__(self, message: Message, context: Context) -> Message:
         """Execute `ClientApp`."""
-        return self._call(message, context)
+        # Execute message using `client_fn`
+        if self._call:
+            return self._call(message, context)
+
+        # Execute message using a new
+        if message.metadata.message_type == MessageType.TRAIN:
+            if self._train:
+                return self._train(message, context)
+            raise ValueError("No `train` function registered")
+        if message.metadata.message_type == MessageType.EVALUATE:
+            if self._evaluate:
+                return self._evaluate(message, context)
+            raise ValueError("No `evaluate` function registered")
+        if message.metadata.message_type == MessageType.QUERY:
```

Comment:
is `MessageType.QUERY` what `get_parameters` should be mapped to?

Reply:
no, `query` is supposed to be a new feature for federated analytics.

we need to decide how we want to proceed with `get_parameters` and `get_properties`.